calico/felix/fiptables.py

```diff
@@ -1,214 +1,451 @@
+        if position == RULE_POSN_LAST:
+            position = len(self.rules)
+
+        if force_position:
+            if (len(self.rules) <= position) or (rule != self.rules[position]):
+                # Either adding after existing rules, or replacing an existing rule.
+                self._insert_rule(rule, position)
+        else:
+            if rule not in self.rules:
+                self._insert_rule(rule, position)
+        return
+
+    def _insert_rule(self, rule, position):
+        """
+        Insert the given rule at the supplied position.
+        
+        position of 0 means "insert at the start", which is what iptables
+        considers as position 1.
+
+        This method is private; callers outside this module should call
+        insert_rule instead.
+        """
+        args = [IPTABLES_CMD[self.type],
+                "-w",
+                "-t",
+                self.table.name,
+                "-I",
+                self.name,
+                str(position + 1)]
+        args.extend(rule.generate_fields())
+        self.table.ops.append(args)
+        self.rules.insert(position, rule)
+        log.debug("Creating rule : %s", args)
+
+    def __str__(self):
+        output = "  Chain %s\n" % self.name
+        for rule in self.rules:
+            output += "    %s\n" % rule
+
+        return output
+
+class Table(object):
+    """
+    Representation of an iptables table.
+    """
+    def __init__(self, type, name):
+        self.type = type
+        self.name = name
+        self.chains = {}
+        self.ops = []
+
+    def is_chain(self, name):
+        return (name in self.chains)
+
+    def get_chain(self, name):
+        """
+        Gets a chain, creating it first if it does not exist.
+        """
+        chain = self.chains.get(name)
+
+        if chain is None:
+            log.debug("Creating chain %s in table %s (%s)",
+                      name,
+                      self.name,
+                      self.type)
+            chain = Chain(self, name)
+            self.ops.append([IPTABLES_CMD[self.type],
+                              "-w",
+                              "-t",
+                              self.name,
+                              "-N",
+                              name])
+
+        return chain
+
+    def delete_chain(self, chain_name):
+        log.debug("Delete chain %s from table %s (%s)",
+                  chain_name,
+                  self.name,
+                  self.type)
+        self.ops.append([IPTABLES_CMD[self.type],
+                        "-w",
+                        "-t",
+                        self.name,
+                        "-X",
+                        chain_name])
+        del self.chains[chain_name]
+
+    def __str__(self):
+        output = "TABLE %s (%s)\n" % (self.name, self.type)
+        for chain_name in sorted(self.chains.keys()):
+            output += str(self.chains[chain_name])
+            output += "\n"
+        return output
```

Comment:
Sorting is useful because we want the output to be deterministic (test outputs tend to rely on it being consistent within a run), and the same as what you get from iptables itself. I would expect the sorting effort (sorting on some short strings) to be very minor, especially for code that isn't called in the mainline.

Agreed that cutting down the concatenations is worthwhile; done that.